lxd/storage/utils.go

```diff
@@ -946,29 +946,18 @@ func volumeIsUsedByDevice(vol api.StorageVolume, instanceType instancetype.Type,
 		}
 	}
 
-	var volName string
-	var snapName string
-
-	if shared.IsSnapshot(vol.Name) {
-		parts := strings.SplitN(vol.Name, shared.SnapshotDelimiter, 2)
-		volName, snapName = parts[0], parts[1]
-	} else if dev["source.snapshot"] != "" {
+	volName, snapName, isSnap := api.GetParentAndSnapshotName(vol.Name)
+	if !isSnap && filters.IsCustomVolumeDiskSnapshot(dev) {
 		// vol is not a snapshot but dev refers to one
 		return false, nil
-	} else {
-		volName = vol.Name
 	}
 
-	volumeTypeName := cluster.StoragePoolVolumeTypeNameCustom
-	if dev["source.type"] != "" {
-		volumeTypeName = dev["source.type"]
+	volumeTypeName := dev["source.type"]
+	if volumeTypeName == "" {
+		volumeTypeName = cluster.StoragePoolVolumeTypeNameCustom
 	}
 
-	if volumeTypeName == vol.Type && dev["source"] == volName && dev["source.snapshot"] == snapName {
-		return true, nil
-	}
-
-	return false, nil
+	return volumeTypeName == vol.Type && dev["source"] == volName && dev["source.snapshot"] == snapName, nil
```

Comment:
For the record I'm not a big fan of this style of moving logic into the final return statement because ive found it complicates refactors in the future. Rather than having a if stanza with a return for each stage of the function, the last part is "special" and can lead to mistakes when refactoring. Not a blocker.